ui/StatusQ/src/concatmodel.cpp

```diff
@@ -0,0 +1,681 @@
+        expectedRoles: ["key", "name"]
+    }
+    \endqml
+*/
+
+SourceModel::SourceModel(QObject* parent)
+    : QObject{parent}
+{
+}
+
+void SourceModel::setModel(QAbstractItemModel* model)
+{
+    if (m_model == model)
+        return;
+
+    if (model)
+        connect(model, &QObject::destroyed, this, &SourceModel::onModelDestroyed);
+
+    if (m_model)
+        disconnect(m_model, &QObject::destroyed, this, &SourceModel::onModelDestroyed);
+
+    emit modelAboutToBeChanged();
+    m_model = model;
+    emit modelChanged(false);
+}
+
+
+/*!
+    \qmlproperty any StatusQ::SourceModel::model
+
+    The model that will be concatenated with other models within ConcatModel.
+*/
+QAbstractItemModel* SourceModel::model() const
+{
+    return m_model;
+}
+
+void SourceModel::setMarkerRoleValue(const QString& markerRoleValue)
+{
+    if (m_markerRoleValue == markerRoleValue)
+        return;
+
+    m_markerRoleValue = markerRoleValue;
+    emit markerRoleValueChanged();
+}
+
+/*!
+    \qmlproperty string StatusQ::SourceModel::markerRoleValue
+
+    The value that will be exposed from the ConcatModel through the role named
+    according to \l {ConcatModel::markerRoleName} for the entries coming from
+    the model defined in SourceModel::model.
+*/
+const QString& SourceModel::markerRoleValue() const
+{
+    return m_markerRoleValue;
+}
+
+void SourceModel::onModelDestroyed()
+{
+    m_model = nullptr;
+    emit modelChanged(true);
+}
+
+
+ConcatModel::ConcatModel(QObject* parent)
+    : QAbstractListModel{parent}
+{
+}
+
+/*!
+    \qmlproperty list<SourceModel> StatusQ::ConcatModel::sources
+
+    This property holds the list of \l {SourceModel} wrappers. Every wrapper
+    holds model which is intended to be concatenated with others within the
+    proxy.
+*/
+QQmlListProperty<SourceModel> ConcatModel::sources()
+{
+    QQmlListProperty<SourceModel> listProperty(this, &m_sources);
+
+    listProperty.replace = nullptr;
+    listProperty.clear = nullptr;
+    listProperty.removeLast = nullptr;
+
+    listProperty.append = [](auto listProperty, auto element) {
+        ConcatModel* model = qobject_cast<ConcatModel*>(listProperty->object);
+
+        if (model->m_initialized) {
+            qWarning() << "Adding sources dynamically is not supported.";
+            return;
+        }
+
+        model->m_sources.append(element);
+    };
+
+    return listProperty;
+}
+
+void ConcatModel::setMarkerRoleName(const QByteArray& markerRoleName)
+{
+    if (m_markerRoleName == markerRoleName)
+        return;
+
+    if (m_markerRoleName != s_defaultMarkerRoleName || m_initialized) {
+        qWarning() << "Property \"markerRoleName\" is intended to be "
+                      "initialized once before roles initialization and not "
+                      "modified later.";
+        return;
+    }
+
+    m_markerRoleName = markerRoleName;
+    emit markerRoleNameChanged();
+}
+
+/*!
+    \qmlproperty string StatusQ::ConcatModel::sources
```

Comment:
```suggestion
    \qmlproperty string StatusQ::ConcatModel::markerRoleName
```

Reply:
Also... the return type won't be a `string` but rather an `array[Char]`; I'd suggest really returning a `QString` here